### PR TITLE
RFC: Simplify decimal (#2440)

### LINKED ISSUE #2440: There is no compiler error when using an invalid Decimal type.
**Describe the bug**
Users can create Decimal values with invalid BYTE_WIDTH, such as
```rust
let v = BasicDecimal::<1>::new(1, 1, &[1]);
```

**Expected behavior**
We need to add more restriction to the decimal type so that only `Decimal128` and `Decimal256` can be used.

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2440

# Rationale for this change
 

Whilst reviewing #2439 I felt something was a bit off, this was my quick attempt to simplify things. **If we are going to do this, we should probably get it in before the release on Thursday**.

# What changes are included in this PR?

Rather than using const generic, we use named type structs. This is the same approach we use elsewhere in the codebase, with Int32Type, etc...

# Are there any user-facing changes?

Yes, this changes the low-level details of how decimal arrays are handled

### CODE REVIEW

File: arrow/src/array/iterator.rs [hunk 1]
```diff
@@ -104,69 +105,14 @@ pub type GenericStringIter<'a, T> = ArrayIter<&'a GenericStringArray<T>>;
...
 /// [`super::Decimal256Array`]
-pub type Decimal256Iter<'a> = BasicDecimalIter<'a, 32>;
-/// an iterator that returns `Some(i128)` or `None`, that can be used on a
-/// [`Decimal128Array`]
-#[derive(Debug)]
-#[deprecated(note = "Please use `Decimal128Iter` instead. \
-    `DecimalIter` iterates `Decimal128` values as i128 values. \
-    This is kept mostly for back-compatibility purpose. Suggests to use `Decimal128Array.iter()` \
-    that returns `Decimal128Iter`.")]
-pub struct DecimalIter<'a> {
```
Comment (tustvold) [inline arrow/src/array/iterator.rs:123]: I wanted to reuse this name, as `BasicDecimal` is confusing. It isn't vital we remove it, but I think it is better to not be stuck on a stale name

File: arrow/src/util/decimal.rs [hunk 2]
```diff
@@ -18,50 +18,26 @@
...
+pub struct Decimal<T: DecimalType> {
     precision: usize,
     scale: usize,
-    value: [u8; BYTE_WIDTH],
+    value: T::Native,
 }
 
-impl<const BYTE_WIDTH: usize> BasicDecimal<BYTE_WIDTH> {
-    #[allow(clippy::type_complexity)]
-    const MAX_PRECISION_SCALE_CONSTRUCTOR_DEFAULT_TYPE: (
```
Comment (tustvold) [inline arrow/src/util/decimal.rs:38]: This was the main thing that felt a bit off, we constrain the permitted constants here, but then also seal the trait in #2439. The whole thing just felt a little esoteric and hard to follow

Comment (HaoYang670) [reply arrow/src/util/decimal.rs:38]: The weird thing I find is that, the rust compiler does the constant evaluation lazily. For example,
```rust
BasicDecimal<1>::try_new(...);
```
will cause a compiler error: "invalid byte length" because the constants we defined in the `impl BasicDecimal` are used in the `try_new` function. However, hackers can run
```rust
BasicDecimal<1>::new(...);
```
without neither compiler error nor runtime error, because no constant is used in this method, so the compiler will not evaluate these constant.

In a word, a hacker can successfully use an invalid decimal type as long as they never touch the constants defined in the `impl BasicDecimal`.

Comment (tustvold) [reply arrow/src/util/decimal.rs:38]: Agreed, this is part of what makes me feel that perhaps using const generics in this way doesn't provide the best UX

Comment (HaoYang670) [reply arrow/src/util/decimal.rs:38]: I think the truth is that, the rust compiler compiles all generic types lazily.
It will only compile the types and methods that are used to avoid large binary output.

**Comment by HaoYang670** [comment]
Personally, I think what makes implementing Decimal elegantly a hard thing is that the behavior of `Decimal` is in the middle of `trait bound` and `generic type`:
```
Trait (Different structure, same behaviour).
/\
 |
\/
Decimal128/256  (very similar structure, but not similar behaviour)  
/\
 |
\/
Generic Type (silmilar structure, similar behaviour)
```
So far, a consensus is that whatever the choice is, we have to implement some part of Decimal128 and Decimal256 separately (although macro can help to simplify this). (For example, the validation function) As currently, Decimal128 is binding to `i128` and Decimal256 is binding to `[u8; 32]` or `BigInt`.

**Comment by HaoYang670** [comment]
Thank you, @tustvold, I am learning your implementation carefully 😁.

File: arrow/src/array/array_decimal.rs [hunk 3]
```diff
@@ -68,24 +71,24 @@ use crate::util::decimal::{BasicDecimal, Decimal256};
 ///    assert_eq!(6, decimal_array.scale());
 /// ```
 ///
-pub type Decimal128Array = BasicDecimalArray<16>;
+pub type Decimal128Array = DecimalArray<Decimal128Type>;
```
Comment (alamb) [inline arrow/src/array/array_decimal.rs:74]: This is certainly a more consistent with the other Array types

**Review by alamb — COMMENTED** [review]
I think this PR makes a of sense to me, but I haven't spent a lot of time in the `DecimalArray` area and I don't really have a strong usecase personally or professionally for this code.

Thus I would defer to @viirya @liukun4515  and @HaoYang670  who have worked in this area more recently or have more directly usecases

**Comment by liukun4515** [comment]
I will review this today later

File: arrow/src/datatypes/types.rs [hunk 4]
```diff
@@ -455,6 +459,68 @@ impl Date64Type {
     }
 }
 
+mod private {
```
Comment (liukun4515) [inline arrow/src/datatypes/types.rs:462]: 👍, use this method to seal the decimal data type.
Other type can implement the `DecimalType`

File: arrow/src/array/mod.rs [hunk 5]
```diff
@@ -199,15 +199,12 @@ pub(crate) use self::data::BufferSpec;
 pub use self::array_binary::BinaryArray;
 pub use self::array_binary::LargeBinaryArray;
 pub use self::array_boolean::BooleanArray;
-pub use self::array_decimal::BasicDecimalArray;
 pub use self::array_decimal::Decimal128Array;
 pub use self::array_decimal::Decimal256Array;
+pub use self::array_decimal::DecimalArray;
```
Comment (liukun4515) [inline arrow/src/array/mod.rs:204]: Do we need add this 
```
#[deprecated(note = "Please use `Decimal128Array` instead")]
```
？

Comment (tustvold) [reply arrow/src/array/mod.rs:204]: No, this is the generic representation of the array, akin to `PrimitiveArray`

File: arrow/src/datatypes/types.rs (hunk 4 above)
Comment (tustvold) [reply arrow/src/datatypes/types.rs:462]: Is this a suggestion or a statement?

Comment (liukun4515) [reply arrow/src/datatypes/types.rs:462]: just statement not suggestion

**Review by liukun4515 — APPROVED** [review]
LGTM
It is a great PR.
Thanks, @tustvold

**Comment by liukun4515** [comment]
If this merged, I can go on this https://github.com/apache/arrow-rs/pull/2357 work

**Review by HaoYang670 — APPROVED** [review]
LGTM.
I couldn't find even a better way to implement `Decimal`, but you have done it @tustvold. Amazing work.

Just a nit. 
Maybe we could add more docs to explain the relationship between `Decimal`, `DecimalType`, `Decimal128Type`, `Decimal256Type` and `NativeDecimalType`, so that other developers can take less time to understand the code.

**Review by viirya — APPROVED** [review]
Looks a more consistent way to align with existing ones like PrimitiveArray. Nice move.

File: arrow/src/array/mod.rs (hunk 5 above)
Comment (viirya) [reply arrow/src/array/mod.rs:204]: 👍

File: arrow/src/array/array_decimal.rs [hunk 6]
```diff
@@ -401,15 +402,15 @@ impl Decimal256Array {
...
+impl<T: DecimalType> From<ArrayData> for DecimalArray<T> {
     fn from(data: ArrayData) -> Self {
         assert_eq!(
             data.buffers().len(),
             1,
             "DecimalArray data should contain 1 buffer only (values)"
         );
         let values = data.buffers()[0].as_ptr();
-        let (precision, scale) = match (data.data_type(), BYTE_WIDTH) {
+        let (precision, scale) = match (data.data_type(), T::BYTE_LENGTH) {
```
Comment (viirya) [inline arrow/src/array/array_decimal.rs:413]: nit:
```suggestion
        let (precision, scale) = match (data.data_type(), Self::VALUE_LENGTH) {
```

**Comment by tustvold** [comment]
Thank you all for reviewing :smile:
